### PR TITLE
Password Reset model + routes

### PR DESCRIPTION
**Pull request by vidushiMaheshwari**
## Password Reset Functionality

Issue Number(s): #12 

What does this PR change and why?
Creates a new table in MongoDB that focuses on temporarily holding password reset requests. Holds email, expiry (the request will die in 7 days), and unique token which would be used in the password reset URL sent to user. Updates the password from Google Auth if user and request is valid.

### Checklist

- [x] Create the `IPasswordReset` type with the mentioned fields
- [x] Create the `PasswordReset` model with the mentioned fields
- [x] `/api/volunteer/password-reset/create` route throws an error if `email` or `name` is undefined
- [x] `/api/volunteer/password-reset/create` route throws an error if `name` does not match the `name` in the corresponding `User` record
- [x] `/api/volunteer/password-reset/create` route deletes any previous `PasswordReset` record associated with `email`
- [x] `/api/volunteer/password-reset/create` route creates a new `PasswordReset` record associated with `email`. The `expiryDate` should be set to 1 week from the time the request is received
- [x] `/api/volunteer/password-reset/validate` route throws an error if `token` is undefined
- [x] `/api/volunteer/password-reset/validate` route throws an error if no `PasswordReset` record is associated with `token` or if the current date is past the `expiryDate`
- [x] `/api/volunteer/password-reset/validate` route returns true otherwise
- [x] `/api/volunteer/password-reset/update-password` route throws an error if `token` or `password` is undefined
- [x] `/api/volunteer/password-reset/update-password` route throws an error if no `PasswordReset` record is associated with `token` or if the current date is past the `expiryDate`
- [x] `/api/volunteer/password-reset/update-password` route grabs the email from the `PasswordReset` record and deletes the record
- [x] `/api/volunteer/password-reset/update-password` route updates the password for `email` using Firebase admin API

### Critical Changes

- Added an extra database
- Added uuid package

### Related PRs
None

### Testing

Enumerate steps to test the functionality of your ticket. This should include edge cases and testing of error handling, if applicable.

### CODE REVIEW

File: server/firebase/passwordReset.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  return value;
+};
+
+export const updateUserPassword = async (
+  email: string,
+  password: string,
+): Promise<boolean> => {
+  const userRecord = await getUserByEmail(email);
+  if (!userRecord) {
+    throw new Error("user not found");
```
Comment (samarth52) [inline server/firebase/passwordReset.ts]: Can you capitalize "user" to "User"

File: src/app/api/volunteer/auth/password-reset/create/route.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+import APIWrapper from "@server/utils/APIWrapper";
+
+type RequestData = {
+  email: string;
+  name: string;
+};
+
+export const POST = APIWrapper({
+  config: {
+    requireToken: true,
```
Comment (samarth52) [inline src/app/api/volunteer/auth/password-reset/create/route.ts]: This route should not require a token since the user will not be signed in for password reset.

File: src/app/api/volunteer/auth/password-reset/update-password/route.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+import APIWrapper from "@server/utils/APIWrapper";
+
+type RequestData = {
+  token: string;
+  password: string;
+};
+
+export const POST = APIWrapper({
+  config: {
+    requireToken: true,
```
Comment (samarth52) [inline src/app/api/volunteer/auth/password-reset/update-password/route.ts]: This route should not require a token since the user will not be signed in for password reset.

File: src/app/api/volunteer/auth/password-reset/validate/route.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
+import {
+  deletePasswordResetField,
+  getPasswordByToken,
+} from "@server/mongodb/actions/PasswordReset";
+import APIWrapper from "@server/utils/APIWrapper";
+
+export const GET = APIWrapper({
+  config: {
+    requireToken: true,
```
Comment (samarth52) [inline src/app/api/volunteer/auth/password-reset/validate/route.ts]: This route should not require a token since the user will not be signed in for password reset.

File: src/app/api/volunteer/auth/password-reset/update-password/route.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    const passwordDoc: IPasswordReset | null = await getPasswordByToken(
+      requestData.token,
+    );
+
+    if (!passwordDoc) {
+      throw new Error("Password reset record is not found");
+    }
+
```
Comment (samarth52) [inline src/app/api/volunteer/auth/password-reset/update-password/route.ts:35]: An error should also be thrown here if the password reset record is expired. Delete the record as well

File: src/app/api/volunteer/auth/password-reset/validate/route.ts [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  handler: async (req) => {
+    const { searchParams } = new URL(req.url);
+    const token = searchParams.get("token");
+    if (!token) {
+      throw new Error("Token parameter is missing in the request.");
+    }
+    const passwordResetField = await getPasswordByToken(token);
+    const currDate = new Date();
+    if (passwordResetField === null) {
+      throw new Error("Password is pasy its expiry");
```
Comment (samarth52) [inline src/app/api/volunteer/auth/password-reset/validate/route.ts]: This message should be "Password reset record was not found".

File: src/app/api/volunteer/auth/password-reset/validate/route.ts [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    }
+    if (passwordResetField.expiryDate < currDate) {
+      await deletePasswordResetField(passwordResetField);
```
Comment (samarth52) [inline src/app/api/volunteer/auth/password-reset/validate/route.ts]: Small change but this condition should be `<=` instead of `<` since `expiryDate` marks the first second where this record is invalid. Also, throw an error "Password reset token has expired."

File: src/app/api/volunteer/auth/password-reset/validate/route.ts [hunk 8]
```diff
@@ -0,0 +1,28 @@
+import {
+  deletePasswordResetField,
+  getPasswordByToken,
+} from "@server/mongodb/actions/PasswordReset";
+import APIWrapper from "@server/utils/APIWrapper";
+
```
Comment (samarth52) [inline src/app/api/volunteer/auth/password-reset/validate/route.ts:7]: Add `export const dynamic = "force-dynamic";` here. The absence of this line throws a "Dynamic server usage" error in production for GET requests that take in search parameters.

**Review by samarth52 — CHANGES_REQUESTED** [review]
Great job on this PR! I have left a couple of comments for some minor changes to be made. Be sure to capitalize your error messages, and make sure to not use the `requireToken` config field of the APIWrapper when the route does not require authentication.

**Review by samarth52 — APPROVED** [review]
lgtm 👍 Amazing job on this PR
